Events Tab with full functionality

cstock123 opened the pull request:
Includes simplified transition and long press gesture. Facebook button, add to calendar button, and location button all works on the Event details screen. Attached is a screenshot of both the cell layout and the events Details screen. Feedback and critiques encouraged on all UI design of the events Section.

![screen shot 2018-07-24 at 1 59 39 pm](https://user-images.githubusercontent.com/32626648/43163275-ebcd58d4-8f4a-11e8-802c-bab521dbc598.png)
![screen shot 2018-07-24 at 2 00 09 pm](https://user-images.githubusercontent.com/32626648/43163285-f051a4d2-8f4a-11e8-9ad4-67c4ab2a852d.png)

cstock123 committed:
added functionality to add to calendar button. Does not work if no EndDate is provided

cstock123 committed:
cleaned up the transition a bit

cstock123 committed:
added location to the cells. fixed calendar functionality for EndDate. Also moved around a few constraints

cstock123 committed:
added functinality to the facebook button. also started to add dynamic height to the cells

cstock123 committed:
fixed small auto height issue

cstock123 committed:
removed legacy code

cstock123 committed:
added alert when add to Calendar button succesfully adds an event to the calendar

cstock123 committed:
started working on the location link of the detail view for events

cstock123 committed:
added functionality to the location button on event detail screen

cstock123 committed:
simplified the gesture recognizer until I can get it working properly

cstock123 committed:
added dynamic height to cells. Still a little glitchy

cstock123 committed:
fixed one more issue with dynamic height of cells

cstock123 committed:
fixed one more bug with dynamic cell height Title Lable

cstock123 commented:
Also, I have lots of gesture code commented out for the time being, but it will be used in the future.

tydahlwave commented on CruCentralCoast/Database/Event.swift:
```diff
@@ -12,26 +12,44 @@ import Firebase
+    var endDate: Date
     var summary: String
-    var location: String?
+    var locationDict: NSDictionary
```
Consider creating a struct to hold the location data. Methods could also be added to the struct to allow easier manipulation of the data. Although there might already exist a data type that would suffice

cstock123 replied:
leaving this for now until we get your code @tydahlwave code working with all of the events. You mentioned over the phone this may not be necessary for now

tydahlwave commented on CruCentralCoast/Database/Event.swift:
```diff
@@ -12,26 +12,44 @@ import Firebase
+        self.locationDict = locationDict
+        self.facebookURL = facebookURL
+        
+        guard let street = locationDict.value(forKey: "street1") as? String,
```
You can access an NSDictionary the same way as a Dictionary, through the [] subscript

tydahlwave commented on CruCentralCoast/Events/EventDetailsVC.swift:
```diff
@@ -7,6 +7,9 @@
 
 import UIKit
+import EventKit //used for the add to calendar button
+import SafariServices //used for the facebookButton
```
Nice comments 👍

tydahlwave commented on CruCentralCoast/Events/EventDetailsVC.swift:
```diff
@@ -49,44 +57,113 @@ class EventDetailsVC: UIViewController {
+    
+    @IBAction func locationButtonPressed(_ sender: Any) {
+        guard let eventLocation = self.eventLocation else {
+            print("error1")
```
Might want a more descriptive error message

hungrylittlebear replied:
I'd try to only use print statements for debugging purposes and only push assertionFailures and console logging

tydahlwave commented on CruCentralCoast/Events/EventDetailsVC.swift:
```diff
@@ -49,44 +57,113 @@ class EventDetailsVC: UIViewController {
+        }
+    }
+    
+//    func getCoordinate(eventLocation: String) -> (CLLocationDegrees, CLLocationDegrees){
```
Remove this

tydahlwave commented on CruCentralCoast/Events/EventDetailsVC.swift:
```diff
@@ -49,44 +57,113 @@ class EventDetailsVC: UIViewController {
+                event.calendar = eventStore.defaultCalendarForNewEvents
+                do {
+                    try eventStore.save(event,span: .thisEvent)
+                        let alert = UIAlertController(title: "Calendar", message: "Event Successfully added to calendar", preferredStyle: .alert)
```
You can also use the presentAlert() method that we added as an extension to reduce these 3 lines to 1 line

cstock123 replied:
very helpful thanks!

tydahlwave commented on CruCentralCoast/Events/EventDetailsVC.swift:
```diff
@@ -49,44 +57,113 @@ class EventDetailsVC: UIViewController {
+                }
+                print("Save Event")
+            } else {
+                print("error : \(String(describing: error))")
```
I don't think you need the String(describing: error) part. It should automatically do that when printing an error

tydahlwave commented on CruCentralCoast/Events/EventDetailsVC.swift:
```diff
@@ -16,21 +19,26 @@ struct EventCellParameters {
     
     @IBOutlet weak var topConstraint: NSLayoutConstraint!
     
     var eventTitle: String?
```
Are these variables needed if the EventDetailsVC already contains a reference to the Event?

tydahlwave commented on CruCentralCoast/Events/EventDetailsVC.swift:
```diff
@@ -7,6 +7,9 @@
+import SafariServices //used for the facebookButton
+import MapKit
 
 struct EventCellParameters {
```
Is this being used? Maybe we don't need it

tydahlwave commented on CruCentralCoast/Events/EventDetailsVC.swift:
```diff
@@ -49,44 +57,113 @@ class EventDetailsVC: UIViewController {
+        }
     }
     
     func configure(with cellParameters: Event) {
```
This function would not be needed if the variables on this class are removed since Event already contains the same info

tydahlwave commented on CruCentralCoast/Events/EventsVC.swift:
```diff
@@ -48,47 +57,102 @@ class EventsTVC: UITableViewController {
         }
     }
+    
+    func longPressGestureBegan(indexPath: IndexPath) {
```
It would be great to consolidate all of these gesture methods in an extension of the class `extension EventsVC: UIGestureRecognizerDelegate`

tydahlwave commented on CruCentralCoast/Events/EventsVC.swift:
```diff
@@ -8,9 +8,13 @@
 import UIKit
 
-class EventsTVC: UITableViewController {
+class EventsTVC: UITableViewController, UIGestureRecognizerDelegate {
```
Can we rename this to EventsVC?

tydahlwave commented on CruCentralCoast/Events/EventsVC.swift:
```diff
@@ -48,47 +57,102 @@ class EventsTVC: UITableViewController {
-        
-        let longPressGestureRecognizer = UILongPressGestureRecognizer(target: self, action: #selector(handleLongPressGesture(gestureRecognizer:)))
-        longPressGestureRecognizer.minimumPressDuration = 0.3
+        createLongPressGesture()
```
Nice use of function decomposition rather than putting all of the code in one giant viewDidLoad method 👍

hungrylittlebear replied:
`self.`

tydahlwave commented on CruCentralCoast/Events/EventsVC.swift:
```diff
@@ -48,47 +57,102 @@ class EventsTVC: UITableViewController {
+    
+    func longPressGestureEnded(indexPath: IndexPath) {
+        let cell = tableView.cellForRow(at: indexPath)
+        guard let vc = UIStoryboard(name: "Events", bundle: nil).instantiateViewController(withIdentifier: "EventDetailsVC") as? EventDetailsVC else {
```
Might want to use the instantiateViewController(_ classType: T.Type) method that we created to make this cleaner

hungrylittlebear commented:
Here's a really quick mockup of two options for how you could style the event cards:

![cards](https://user-images.githubusercontent.com/41555194/43354617-948822fe-9204-11e8-882d-0ac1e01a418c.png)

I think what you have is good, I'm just list all of my thoughts, feel free to disregard or follow up for any of these:
- don't have the date and the title be in the same line because you end up word wrapping very early and not utilizing the horizontal space for the most important content in that piece
- never have word wrapping on. just set the text to scale down once it starts to overflow (pretty easy to do, it's just a setting in the .xib)
- have the cards be taller and allocate more space for the image, the ratio of image to detail view feels off
- I'd rather see the location just on the event detail view that way it's always intractable, the colored icon on the card feels out of place ¯\_(ツ)_/¯ 
- since you're going for a very app store style of card, I'd increase the spread on those shadows by a lot and lower the opacity on it. they should be overlapping with the other cards' shadows

hungrylittlebear commented:
really quick mockup again, don't pay attention too much to the font sizes, spacing, etc. just the general layout.
![iphone x copy](https://user-images.githubusercontent.com/41555194/43355004-aa2a8b60-9209-11e8-88f2-599fbf4bae98.png)

general thoughts, again, no pressure to take any of this feedback:
- make sure you have a close button, even if you have pull down to dismiss, people won't always assume that gesture
- follow a readable width, so give yourself some more padding on the margins for the text, try to have at least 20 leading/trailing while in portrait
- make the image taller
- make your buttons styled similarly to the ones on the login screen, I think that code has already been put in a generic place (hopefully) so you can just reuse that.
- more padding for the icons, etc. but using the other button will probably address all of that

hungrylittlebear commented on CruCentralCoast/Events/EventDetailsVC.swift:
```diff
@@ -49,44 +57,113 @@ class EventDetailsVC: UIViewController {
+        self.imageView.image = self.eventImage
+//        self.locationButton.titleLabel?.text = self.eventViewableLocation
+        self.locationButton.setTitle(self.eventViewableLocation, for: .normal)
+//        print(self.locationButton.titleLabel?.text)
```
you can drop the commented out code

hungrylittlebear commented on CruCentralCoast/Events/EventDetailsVC.swift:
```diff
@@ -49,44 +57,113 @@ class EventDetailsVC: UIViewController {
+        let geoCoder = CLGeocoder()
+        geoCoder.geocodeAddressString(eventLocation) { (placemarks, error) in
+            guard let clPlacemark = placemarks?.first
+            else {
```
I'd keep the `else {` on the same line as the guard

hungrylittlebear commented on CruCentralCoast/Events/EventDetailsVC.swift:
```diff
@@ -49,44 +57,113 @@ class EventDetailsVC: UIViewController {
+            config.entersReaderIfAvailable = true
+            
+            let vc = SFSafariViewController(url: url, configuration: config)
+            present(vc, animated: true)
```
`self.`

hungrylittlebear commented on CruCentralCoast/Events/EventDetailsVC.swift:
```diff
@@ -49,44 +57,113 @@ class EventDetailsVC: UIViewController {
             if (granted) && (error == nil) {
-                let event = EKEvent(eventStore: eventStore)
-                event.title = self.titleLabel.text
+                let event:EKEvent = EKEvent(eventStore: eventStore)
```
I'd say be consistent throughout the project with how you place the spaces for these variable declarations. above on 136 you have `eventStore : EKEventStore` and here `event:EKEvent` my team's preferred way is the following:
`let event: EKEvent`

hungrylittlebear commented on CruCentralCoast/Events/EventsVC.swift:
```diff
@@ -48,47 +57,102 @@ class EventsTVC: UITableViewController {
     @objc internal func handleLongPressGesture(gestureRecognizer: UILongPressGestureRecognizer) {
         let touchPoint = gestureRecognizer.location(in: self.view)
         if gestureRecognizer.state == .began {
+            self.longPressCancelled = false
```
we try to avoid adding state variables as much as we can like `longPressCancelled` and `longPressFailed` I'd recommend giving it your best shot to see if you can cover the gesture cancelled case without needing to track them like this.

cstock123 replied:
This is definitely something I want to fix in the future, but as for right now I am having trouble getting it to work. I completely agree, and this will be fixed when the gesture is finalized.

cstock123 committed:
Merge branch 'release_2.0' into Camden/Event

cstock123 committed:
fixed some but not all pull request details

cstock123 committed:
Merge branch 'Camden/Event' of github.com:CruCentralCoast/CruiOS into Camden/Event

cstock123 committed:
more issues fixed

cstock123 commented:
@hungrylittlebear Thank you for the UI feedback. I implemented a lot of those things. 

For the Cell view here are a few of the things I did and some of the things I had trouble with: 
- Extended the title to take up the entire line and moved the date to the line above. This looks a lot better. 
- I turned word wrapping off. I'm not quite sure what you are talking about with the scaleDown property?
- I am having a problem with making the image larger. If I do this, the text on the image gets cut off. Could be a problem with the aspect fill/aspect fit property. Let me know if you have ideas on that.
- Moved the location to only the detail screen and added a short description 1 line text box.
- Lowered the opacity on the shadows and increased the shadow radius.

Detail View Changes:
- I did have a close button before but I changed it to an actual "X" Button
- extended the padding on the sides to 20 pts 
- Same issue as above with increasing the height on the image. 
- The buttons should be pulling from the same buttons that were used for the login screen. I do agree the there needs to be more padding for the icon.

Again, thanks for the feedback. Let me know what you think with these change descriptions.

cstock123 committed:
Some UI changes

cstock123 committed:
Changed event Imgages to aspect ratio 16:9. Produces taller image

cstock123 committed:
removed all animatedTransition Code and fixed one more UI issue

tydahlwave commented:
I'm gonna approve this PR and get it merged for now.
However, I think the design should follow more of what Connor posted above. Another PR will need to be opened once those UI changes have been made.
